[Java] DataInput implementation over DirectBuffer

tkowalcz opened the pull request:
This PR addresses #273. I have added `DirectBufferDataInput` which implements JDK interface `DataInput`. This implementation adheres to the `DataInput` specification which, sadly, includes throwing checked EOF exception on underflow and being big endian by default. User can change byte order to make it read Agrona buffer's generated data. 

As a bonus it has few additional methods to read Agrona style strings into an `Appendable`.

mjpt777 commented on agrona/src/main/java/org/agrona/io/DirectBufferDataInput.java:
```diff
@@ -0,0 +1,437 @@
+        }
+
+        this.buffer = buffer;
+        this.length = length + offset;
```
Validation of offset and length would be good to check against buffer.

tkowalcz commented:
I have committed changes addressing review comments. 

Btw. I'm pretty sure that condition from `UnsafeBuffer::boundsCheckWrap`:

```java
 if ((offset > capacity - length) || (length > capacity - offset))
{
}
```
can be simplified to have just one of these inequalities as they are equivalent.

mjpt777 commented:
You should add validation of all the wrap parameters.

tkowalcz commented:
You are right. Sorry. Pushed now (includes validation).

tkowalcz commented:
I don't think that will work. I'll create second PR for this.
